core/server/data/migrations/versions/2.27/3-add-subdirectory-to-relative-canonical-urls.js

```diff
@@ -0,0 +1,104 @@
+const models = require('../../../../models');
+const common = require('../../../../lib/common');
+const config = require('../../../../config');
+const {URL} = require('url');
+
+module.exports.config = {
+    transaction: true
+};
+
+// We've been incorrectly saving canonical_url fields without a subdirectory.
+// We need this column to be consistent with all other relative URLs, so...
+
+// if we have a configured url with a subdirectory
+// find all posts that have a canonical_url starting with / but not //
+// prefix the subdirectory to the canonical_url and save the post
+module.exports.up = (options) => {
+    // normalize config url to always have a trailing-slash
+    let configUrl = config.get('url');
+    if (!configUrl.endsWith('/')) {
+        configUrl = `${configUrl}/`;
+    }
+    const url = new URL(config.get('url'));
+
+    const localOptions = Object.assign({
+        context: {internal: true}
+    }, options);
+
+    if (url.pathname === '/') {
+        common.logging.info('Skipping posts.canonical_url subdirectory fix: no subdirectory configured');
+        return Promise.resolve();
+    }
+
+    // perform a specific query for the type of canonical URLs we're looking for
+    // so we're not fetching and manually looping over a ton of post models
+    return models.Posts
+        .forge()
+        .query((qb) => {
+            qb.where('canonical_url', 'like', '/%');
+            qb.whereNot('canonical_url', 'like', '//%');
+        })
+        .fetch(localOptions)
+        .then((posts) => {
+            if (posts) {
+                const queries = [];
+
+                posts.forEach((post) => {
+                    const canonicalUrl = post.get('canonical_url').replace('/', url.pathname);
+                    post.set('canonical_url', canonicalUrl);
+                    queries.push(post.save(null, localOptions));
+                });
+
+                return Promise.all(queries).then(() => {
```

Comment:
Think it's more safe to use [Promise.mapSeries](http://bluebirdjs.com/docs/api/promise.mapseries.html) in this situation just to avoid clogging the connection pool in case there are loads of urls found to be migrated. Promise.all resolved all operations at once without waiting in turns unlike mapSeries:
> The mapper returns a promise or a thenable, it is awaited before continuing to the next iteration.

Reply:
Good point, makes the code more concise too! Update pushed